### ISSUE TITLE
Builds are taking too long

### ISSUE THREAD

**Issue by zmerlynn** [kubernetes/kubernetes#27444]
Over a year ago, we had the `kubernetes-build` job down <5m. It's now coming in at 30m. Some of that shot up really recently, too, looking at the build time trend:

![time-trend](https://cloud.githubusercontent.com/assets/4942464/16088152/95720012-32da-11e6-981c-410fa6a2abb1.png)

It looks like sometime around #10384, we shot up from a fat 20m another 50% to 30m. Can we at least eliminate that jump?

Is there anything else we can do here? Maybe split the non-`amd64` builds off to a separate job so that our whole CI infrastructure isn't tied to them?

@kubernetes/test-infra-maintainers @kubernetes/release-maintainers

**Comment by spxtr** [comment]
The huge jump was when we started building on arm and ppc. I think that took us from 20 to 30 min. @luxas knows more.

> Maybe split the non-amd64 builds off to a separate job so that our whole CI infrastructure isn't tied to them?

That sounds like an excellent idea. We can do a quick release for triggering CI and have separate jobs making releases. @david-mcmahon

**Comment by spxtr** [comment]
A related problem is that the size of the output is growing. `make release` will generate a 16G `_output` folder.

**Comment by ixdy** [comment]
@spxtr the jump from 20m to 30m was far more recent - Sunday night. @fejta anything weird happen to Jenkins over the weekend?

**Comment by spxtr** [comment]
Oh interesting. That might be when we switched to 8-core machines?

**Comment by luxas** [comment]
We should **not** remove non-amd64, builds from CI, otherwise it's very likely that releases are going to break every now and then, because we haven't tried to compile it. The main point with having the arches in CI right now is that we will spot compilation errors in an early stage (PRs), and may investigate issues if there are any. Like what @smarterclayton and I did with ppc64le (which is going to be re-enabled soon again)

We may however, if you think it's worthwhile, parallelize the jobs so in e2e we'll do a quick-release and run all tests, and in a cross-build job, we'll make sure that Kubernetes builds on all platforms that are in `make release`. That's the option we have

**Comment by rmmh** [comment]
How often do non-amd64 builds actually fail? We could put them in a separate queue-blocking job (like the e2e jobs) and catch those failures slightly slower in exchange for faster/smaller builds for the general case.

**Comment by gmarek** [comment]
I don't think @zmerlynn suggested removing non-amd64 build from CI, just removing them from the main 'kubernetes-build' which generates binaries used by other tests (which are running on amd64 machines). By all means we should have another 'kubernetes-non-amd64-build' build as a blocker for the merge queue.

**Comment by zmerlynn** [comment]
== @gmarek. I'm proposing a separate build for all arches that can be off-schedule from the rest of the CI infrastructure.

**Comment by zmerlynn** [comment]
(even triggered by the `amd64` build succeeding)

**Comment by ixdy** [comment]
For reference, 1.2 builds take about 10m, and there wasn't an appreciable jump in the most recent build yesterday.

**Comment by ixdy** [comment]
`kubernetes-build` went from about 16m to 21-22m starting at build 10361, right when it appears that agents were rebuilt. there was an outlier at build 10366 which took 27, and then we consistently started taking 30m+ at build 10385. There might be code changes to explain this, though I'm guessing that Jenkins infra changes are more likely to blame.

**Comment by luxas** [comment]
Ah, all good then. Somehow I misinterpreted the two first comments.
Separating amd64 e2e and non-amd64 compilation sounds good to me.

@rmmh I expect non-amd64 builds to fail 0-2 times per minor release, so it doesn't fail often. 
This time, v1.2-v1.3, the ppc64le hyperkube build failed (somehow it has grown too big for the ppc64le gcc compiler or something like that)

**Comment by fejta** [comment]
What does improving this going to gain us given the time required for all the other blocking jobs?

I posit that time would be better spent making serial go faster, for example

**Comment by gmarek** [comment]
I believe it can be done in less time that we already spent on this issue. Someone correct me if I'm wrong.

**Comment by wojtek-t** [comment]
@fejta - this is super important, because it also affects presumbit.
So to clarify - currently e2e tests that are running as part of presubmit takes ~55 minutes. However, out of those we spend:
- 30m of build
- 7m on cluster startup
- 12m on tests
- 7m on cluster teardown

To to be clear - build is consuming more than half of e2e presubmit.
So in fact this is super super important if we want to fix #25940

@mikedanese @lavalamp

**Comment by wojtek-t** [comment]
So this is dependency of P0 issue, so in my opinion this should also be a P0.

**Comment by fejta** [comment]
@wojtek-t the same VMs that run presubmit jobs on each PR are different from the ones that run CI jobs on postsubmit. I agree that speeding up presubmit is a good idea, but it seems orthogonal to how long kubernetes-build takes.

**Comment by wojtek-t** [comment]
>
@wojtek-t the same VMs that run presubmit jobs on each PR are different from the ones that run CI jobs on postsubmit.

I know that. But we are running the same "build" command in both places. And if you look into presubmit e2e tests, first 30-35 minutes of the run is build. And this is far from being acceptable.

**Comment by lavalamp** [comment]
Rules:
* Presubmits must take no longer than 30 minutes. Humans need to be paged and investigating it if it goes about 30 minutes! Our target should actually be like 20 minutes.
* Everything tested in postsubmit should be tested in presubmit (with smaller workloads).

If we held the above two rules our throughput and flake level would be drastically better.

**Comment by spxtr** [comment]
We should probably redesign our PR Jenkins workflow. This seems reasonable:

* A full build/release job
* A quick build that then starts an E2E run on GCE and GKE
* Verification test
* Unit tests
* Integration tests

We've split the verification checks, and I'd like to get a smoke GKE job soon.

**Comment by fejta** [comment]
FYI I added 7 highcpu-16 core executors, will see how they do overnight and if things look good replace all the standard-8 instances with these tomorrow.

**Comment by fejta** [comment]
Results:

[kubernetes-pull-verify-all](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-verify-all/buildTimeTrend):

* ~22m on [standard-8](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-verify-all/437/)
* ~18m on [highcpu-16](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-verify-all/423/)

[kubernetes-pull-test-unit-integration](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-test-unit-integration/buildTimeTrend):

* ~30m on [standard-8](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-test-unit-integration/31056/):
* ~22m on [highcpu-16](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-test-unit-integration/31059/)

[kubernetes-pull-build-test-e2e-gce](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/buildTimeTrend):

* ~35-70m on [standard-8](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/45108/)
* ~45-55m on [highcpu-16](http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/45106/)

I think the biggest issues right now are:

* We can run up to 40 unit/integration instances in parallel but never go more than 12 e2e instances at once
* The e2e jobs are taking the most time and do not seem CPU bound.
* Until we can get the e2e jobs below 30m there isn't much value to running the unit tests faster.

I'm going to try a different experiment -- perhaps the e2e jobs are disk or ram bound instead?

**Comment by wojtek-t** [comment]
@fejta - we know where the problem in e2e-gce is - it is "BUILD":
From a random run:
```
2016/06/16 23:54:18 e2e.go:214: Step 'build-release' finished in 29m17.903395903s
```

**Comment by wojtek-t** [comment]
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/45133/consoleFull

**Comment by wojtek-t** [comment]
Basically, the only thing that we need to do is to set "KUBE_FASTBUILD=true" in "kubernetes-pull-build-test-e2e-gce" and anywhere else for now.

@fejta @spxtr - where are configs for "kubernetes-pull-build-test-e2e-gce" job? I can't find them anywhere.

**Comment by wojtek-t** [comment]
Note that doing this should roughly double our merge throughput

**Comment by spxtr** [comment]
e2e pull configs are in https://github.com/kubernetes/test-infra/pull/169. I'll get that in ASAP. I'm not sure if we should switch on `KUBE_FASTBUILD` until we have a job doing a full release, or else things might break on other OSs.

**Comment by wojtek-t** [comment]
@spxtr - yeah - those would be super useful. (IIUC, so far this was configured in Jenkins itself, right?)

**Comment by luxas** [comment]
This is post-1.3 work, right?
If you then may spin up the jenkins `make release` job, I think it's ok to run just fast-build on e2e

**Comment by wojtek-t** [comment]
This is slightly unrelated to 1.3 - this is purely test-infra related thing. So we may do it in parallel.

**Comment by lavalamp** [comment]
Filed: https://github.com/kubernetes/kubernetes/issues/27578

**Comment by gmarek** [comment]
@spxtr - isn't kubernetes-build building for all architectures? We're already blocking on it.

**Comment by lavalamp** [comment]
You mean we're not already doing `make quick-release`? or is KUBE_FASTBUILD something else?

@gmarek yeah that's the problem statement...

**Comment by spxtr** [comment]
`make quick-release` is the same as `KUBE_FASTBUILD`. We're not using it right now.

**Comment by gmarek** [comment]
So we're testing builds on other architectures in our continuous builds, right? Seeing that the breaks of only non-amd64 builds are pretty rare I don't think we need to run it on every PR if it's taking a lot of time. Can we just use `KUBE_FASTBUILD` for per-PR builds?

**Comment by quinton-hoole** [comment]
+1 !!

**Comment by luxas** [comment]
If it's possible, it would be great with a presubmit jenkins job that does run in parallell with KUBE_FASTBUILD e2e tests.

Otherwise, the SQ will be blocked _after_ a breaking change for non-amd64 builds, not before merging the PR

**Comment by fejta** [comment]
Yes, agreed. We are working on this today -- and getting the e2e PR jobs under source control.

**Comment by gmarek** [comment]
@luxas - it's so rare that I don't think it's a big problem. There's a number of tests that we don't run on each PR, even though they're important - it's just a tradeoff between speed of builds and how quickly we notice a regression. IIUC we have a lot of problems with the first one, while no one is complaining about the second.

**Comment by luxas** [comment]
For me it doesn't actually matter, as long as the multiarch builds doesn't hurt :)
If this happens, we have to revert the PR that did the breaking change, so if _everyone_ agrees with that, I'm ok with it

**Comment by fejta** [comment]
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/27570/kubernetes-pull-build-test-e2e-gce/45359 shows the build takes 7m now, but the overall suite continues to require 50m

```
2016/06/17 18:27:54 e2e.go:214: Step 'build-release' finished in 7m7.01992286s
2016/06/17 18:33:36 e2e.go:214: Step 'teardown' finished in 5m25.440911483s
2016/06/17 18:43:46 e2e.go:214: Step 'up' finished in 9m29.69017303s
2016/06/17 19:00:32 e2e.go:214: Step 'Ginkgo tests' finished in 16m16.739549889s
2016/06/17 19:11:01 e2e.go:214: Step 'teardown' finished in 9m58.537132186s
```

**Comment by fejta** [comment]
Jobs that need to migrate to a new project:

node e2e - https://github.com/kubernetes/kubernetes/issues/27648
e2e - https://github.com/kubernetes/test-infra/issues/187
heapster/cadvisor - https://github.com/kubernetes/test-infra/issues/190

**Comment by wojtek-t** [comment]
@fejta is probably correct, because there are runs that take <30m, e.g. this one:
http://kubekins.dls.corp.google.com:8081/job/kubernetes-pull-build-test-e2e-gce/45838/consoleFull

**Comment by gmarek** [comment]
@luxas - I think that we just live with the possibility of revert caused by a lot of other suites, so I don't think mutliarch build is treated any different.

**Comment by luxas** [comment]
Okay, I'm a huge fan of a faster Submit Queue (I assume everyone agrees that it should be faster), and faster builds, so given that there are other mandatory and blocking tests that aren't a part of the PR checks, I'm fine with moving multiarch.

**Comment by spxtr** [comment]
Now that we have fastbuild on they're taking 30-40 mins, with ~7 min being the build. Can we close this?

**Comment by zmerlynn** [comment]
We're up to 40m now. :(

![screenshot from 2016-06-28 18 58 57](https://cloud.githubusercontent.com/assets/4942464/16438209/69b692fc-3d62-11e6-9828-bda92482f7af.png)

**Comment by zmerlynn** [comment]
I really want the `kubernetes-build` job to be down at manageable levels as well, because it's not un-often tied to humans watching the results (Hi!). Given that it used to be down in the single digits, this is ... :(

**Comment by spxtr** [comment]
I would be okay with the `kubernetes-build` job doing a quick-release, pushing that to whatever CI bucket we're using, and the other jobs using that.

We would need another job (merge-blocking, but that shouldn't be a problem) doing a full cross-platform build.

**Comment by fejta** [comment]
@zmerlynn how about creating a `kubernetes-build` that does a fastbuild and a `kubernetes-build-all` that does everything?

**Comment by zmerlynn** [comment]
Yeah, especially if `kubernetes-build-all` isn't the thing publishing the `latest.txt`.

We could alternately do something crazy and split out `latest.txt` by arch, i.e. `latest-amd64.txt`, but it's already somewhat combinatoric how many latests we publish. :)

**Comment by luxas** [comment]
I don't see a case where latest-amd64 would be anything else than latest-arm...
Why not just do a quick release in the PR e2e test and a full release build somewhere else (in a SQ-blocking test or a pre-commit Jenkins test)

**Comment by zmerlynn** [comment]
@luxas: Because if `kubernetes-build` is fastbuild, the it won't be uploading the binaries for all architectures, so the `latest.txt` file is a bit of a lie in that case.

**Comment by gmarek** [comment]
Can we just have latest-not-amd64.txt specially for this case? I think we should do absolutely everything we can to make it as fast as possible, as long as it doesn't compromise the test coverage in the long run (i.e. as long as we block SQ no more than small X hours after the regression).

**Comment by jbeda** [comment]
If we move to just building hyperkube (and deprecate the individual binaries) we'll have less to build and package up.  I imagine that will speed things up. #16508

**Comment by chrislovecnm** [comment]
👍 currently watching paint dry ...  Working on e2e tests is a tad painful 😦

**Comment by fejta** [comment]
Duplicate of https://github.com/kubernetes/contrib/issues/1522
